kernel/src/checkpoint/mod.rs

```diff
@@ -2,7 +2,613 @@
+//!    include a [`CheckpointMetadata`] action.
+//! 2. **Single-file Classic-named V2 Checkpoint** – for tables supporting the `v2Checkpoints` feature.
+//!    These checkpoints follow the V2 specification and include a [`CheckpointMetadata`] action, while
+//!    maintaining backwards compatibility by using classic-naming that legacy readers can recognize.
+//!
+//! For more information on the V1/V2 specifications, see the following protocol section:
+//! <https://github.com/delta-io/delta/blob/master/PROTOCOL.md#checkpoint-specs>
+//!
+//! ## Checkpoint Selection Logic
+//! The checkpoint type is determined by whether the table supports the `v2Checkpoints` reader/writer feature:
+//!
+//! | Table Feature    | Resulting Checkpoint Type     |
+//! |------------------|-------------------------------|
+//! | No v2Checkpoints | Single-file Classic-named V1  |
+//! | v2Checkpoints    | Single-file Classic-named V2  |
+//!
+//! ## Architecture
+//!
+//! - [`CheckpointWriter`] - Core component that manages checkpoint creation workflow
+//! - [`CheckpointData`] - Contains the data to write and destination path information
+//! - [`crate::log_replay`] submodule - Handles action filtering and deduplication
+//!
+//! ## [`CheckpointWriter`]
+//! Handles the actual checkpoint data generation and writing process. It is created via the
+//! [`crate::table::Table::checkpoint`] method and provides the following APIs:
+//! - [`CheckpointWriter::checkpoint_data`] - Returns the checkpoint data and path information
+//! - [`CheckpointWriter::finalize`] - Writes the `_last_checkpoint` file
+//!
+//! ## Example: Writing a classic-named V1/V2 checkpoint (depending on `v2Checkpoints` feature support)
+//!
+//! TODO(seb): unignore example
+//! ```ignore
+//! let path = "./tests/data/app-txn-no-checkpoint";
+//! let engine = Arc::new(SyncEngine::new());
+//! let table = Table::try_from_uri(path)?;
+//!
+//! // Create a checkpoint writer for the table at a specific version
+//! let mut writer = table.checkpoint(&engine, Some(2))?;
+//!
+//! // Retrieve checkpoint data
+//! let checkpoint_data = writer.checkpoint_data()?;
+//!
+//! // Write checkpoint data to storage (implementation-specific)
+//! let metadata = your_storage_implementation.write_checkpoint(
+//!     &checkpoint_data.path,
+//!     checkpoint_data.data
+//! )?;
+//!
+//! /* IMPORTANT: All data must be written before finalizing the checkpoint */
+//!
+//! // Finalize the checkpoint by writing the _last_checkpoint file
+//! writer.finalize(&engine, &metadata)?;
+//! ```
+//! ## Future extensions
+//! - Single-file UUID-named V2 checkpoints (using `n.checkpoint.u.{json/parquet}` naming) are to be
+//!   implemented in the future. The current implementation only supports classic-named V2 checkpoints.
+//! - Multi-file V2 checkpoints are not supported yet. The API is designed to be extensible for future
+//!   multi-file support, but the current implementation only supports single-file checkpoints.
+//!
```

Comment:
I'd like to see github issues and TODOs associated with these so that they're tracked. Something like: 

- TODO(433):  Support Single-file

@zachschuermann This is a pattern I've seen in other repositories, and I'd advocate for this to be the way we track todos.